packages/visual-editor/src/components/Draggable/Placeholder.tsx

```diff
@@ -0,0 +1,180 @@
+
+  const dropzoneElement = document.querySelector(`[data-rfd-droppable-id="${dropzoneElementId}"]`);
+  const element = document.querySelector(`[data-ctfl-draggable-id="${id}"]`);
+
+  if (!dropzoneElement || !element) {
+    return emptyStyles;
+  }
+
+  const { width, height } = dropzoneElement.getBoundingClientRect();
+
+  if (direction === 'horizontal') {
+    const top = (height - element.getBoundingClientRect().height) / 2;
+
+    if (isEnd) {
+      return { width: DRAGGABLE_WIDTH, height, top: -top, right: -DRAGGABLE_WIDTH };
+    }
+
+    return { width: DRAGGABLE_WIDTH, height, top: -top, left: -DRAGGABLE_WIDTH };
+  }
+
+  const left = (width - element.getBoundingClientRect().width) / 2;
+  const top = -DRAGGABLE_HEIGHT;
+
+  if (isEnd) {
+    return { width, height: DRAGGABLE_HEIGHT, bottom: -DRAGGABLE_HEIGHT, left: -left };
+  }
+
+  return { width, height: DRAGGABLE_HEIGHT, top, left: -left };
+};
+
+const calcMovementStyles = (params: CalcStylesParams): CSSProperties => {
+  const {
+    destinationIndex,
+    sourceIndex,
+    destinationId,
+    sourceId,
+    elementIndex,
+    dropzoneElementId,
+    id,
+    direction,
+    totalIndexes,
+    draggableId,
+  } = params;
+
+  const isEnd = destinationIndex === totalIndexes && elementIndex === totalIndexes - 1;
+  const isEmptyZone = totalIndexes === 0;
+
+  const isSameZone = destinationId === sourceId;
+  const dropzoneElement = document.querySelector(`[data-rfd-droppable-id="${dropzoneElementId}"]`);
+  const draggableElement = document.querySelector(`[data-rfd-draggable-id="${draggableId}"]`);
+  const element = document.querySelector(`[data-ctfl-draggable-id="${id}"]`);
+
+  if (!dropzoneElement || !element || !draggableElement) {
+    return emptyStyles;
+  }
+
+  const dropzoneSizes = dropzoneElement.getBoundingClientRect();
+
+  const draggableSizes = draggableElement.getBoundingClientRect();
+  const elementSizes = element.getBoundingClientRect();
+
+  if (direction === 'horizontal') {
+    const width = draggableSizes.width;
+    const height = dropzoneSizes.height;
+    const top = (dropzoneSizes.height - elementSizes.height) / 2;
+
+    if (isSameZone && destinationIndex > sourceIndex) {
+      return { width, height, top: -top, right: -width };
+    }
+    if (isEnd) {
+      return { width, height, top: -top, right: -width };
+    }
+
+    return { width, height, top: -top, left: -width };
+  }
+
+  const width = dropzoneSizes.width;
+  const height = draggableSizes.height;
+  const top = -height;
+  const left = (dropzoneSizes.width - elementSizes.width) / 2;
+
+  if (isEmptyZone) {
+    return { width, height, bottom: -height, left: -left };
+  }
+  if (isSameZone && destinationIndex > sourceIndex) {
+    return { width, height, bottom: -height, left: -left };
+  }
+
+  if (isEnd) {
+    return { width, height, bottom: -height, left: -left };
+  }
+
+  return { width, height, top, left: -left };
```

Comment:
I notice the returned objects are very similar in each branch here, so it could possibly be refactored to have a single return statement and conditional properties in the object like this:

```ts
const isNegativeTop = direction === 'horizontal' && (isSameZone && destinationIndex > sourceIndex || isEnd);
const isNegativeLeft = direction !== 'horizontal' && (isEmptyZone || isSameZone && destinationIndex > sourceIndex || isEnd);

const width = direction === 'horizontal' ? draggableSizes.width : dropzoneSizes.width;
const height = direction === 'horizontal' ? dropzoneSizes.height : draggableSizes.height;
const top = direction === 'horizontal' ? (dropzoneSizes.height - elementSizes.height) / 2 : -height;
const left = direction === 'horizontal' ? -width : (dropzoneSizes.width - elementSizes.width) / 2;

return {
  width,
  height,
  top: isNegativeTop ? -top : top,
  right: direction === 'horizontal' ? -width : undefined,
  bottom: direction !== 'horizontal' ? -height : undefined,
  left: isNegativeLeft ? -left : left,
};
```

Reply:
Cleannnnn 🔥